lib/bpf/bpf.go

```diff
@@ -0,0 +1,515 @@
+	}
+
+	// Stop watching for events from this PID.
+	s.removeWatch(cgroupID)
+
+	// Move all PIDs to the root cgroup and remove the cgroup created for this
+	// session.
+	err = s.cgroup.Remove(ctx.SessionID)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// loop pulls events off the perf ring buffer, parses them, and emits them to
+// the audit log.
+func (s *Service) loop() {
+	for {
+		select {
+		// Program execution.
+		case event := <-s.exec.events():
+			s.emitCommandEvent(event)
+		// Disk access.
+		case event := <-s.open.events():
+			s.emitDiskEvent(event)
+		// Network access (IPv4).
+		case event := <-s.conn.v4Events():
+			s.emit4NetworkEvent(event)
+		// Network access (IPv4).
+		case event := <-s.conn.v6Events():
+			s.emit6NetworkEvent(event)
+		case <-s.closeContext.Done():
+			return
+		}
+	}
+}
+
+// emitCommandEvent will parse and emit command events to the Audit Log.
+func (s *Service) emitCommandEvent(eventBytes []byte) {
+	var event rawExecEvent
+
+	err := binary.Read(bytes.NewBuffer(eventBytes), bcc.GetHostByteOrder(), &event)
+	if err != nil {
+		log.Debugf("Failed to read binary data: %v.", err)
+		return
+	}
+
+	// If the event comes from a unmonitored process/cgroup, don't process it.
+	ctx, ok := s.watch[event.CgroupID]
+	if !ok {
+		return
+	}
+
+	// If the command event is not being monitored, don't process it.
+	_, ok = ctx.Events[teleport.EnhancedRecordingCommand]
+	if !ok {
+		return
+	}
+
+	switch event.Type {
+	// Args are sent in their own event by execsnoop to save stack space. Store
+	// the args in a ttlmap so they can be retrieved when the return event arrives.
+	case eventArg:
+		var buf []string
+		buffer, ok := s.argsCache.Get(strconv.FormatUint(event.PID, 10))
+		if !ok {
+			buf = make([]string, 0)
+		} else {
+			buf = buffer.([]string)
+		}
+
+		argv := (*C.char)(unsafe.Pointer(&event.Argv))
+		buf = append(buf, C.GoString(argv))
+		s.argsCache.Set(strconv.FormatUint(event.PID, 10), buf, 24*time.Hour)
+	// The event has returned, emit the fully parsed event.
+	case eventRet:
+		// The args should have come in a previous event, find them by PID.
+		args, ok := s.argsCache.Get(strconv.FormatUint(event.PID, 10))
+		if !ok {
+			log.Debugf("Got event with missing args: skipping.")
+			lostCommandEvents.Add(float64(1))
+			return
+		}
+		argv := args.([]string)
+
+		// Convert C string that holds the command name into a Go string.
+		command := C.GoString((*C.char)(unsafe.Pointer(&event.Command)))
+
+		// Emit "command" event.
+		eventFields := events.EventFields{
+			// Common fields.
+			events.EventNamespace:  ctx.Namespace,
+			events.SessionEventID:  ctx.SessionID,
+			events.SessionServerID: ctx.ServerID,
+			events.EventLogin:      ctx.Login,
+			events.EventUser:       ctx.User,
+			// Exec fields.
+			events.PID:        event.PPID,
+			events.PPID:       event.PID,
+			events.CgroupID:   event.CgroupID,
+			events.Program:    command,
+			events.Path:       argv[0],
```

Comment:
Does `events.Path` contain the command path or will it be empty (where path is part of the `program`) ?